File: src/python/grpcio_tests/tests/qps/histogram.py
# Copyright 2016 gRPC authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math
import threading

from src.proto.grpc.testing import stats_pb2


class Histogram(object):
    """Histogram class used for recording performance testing data.

  This class is thread safe.
  """

    def __init__(self, resolution, max_possible):
        self._lock = threading.Lock()
        self._resolution = resolution
        self._max_possible = max_possible
        self._sum = 0
        self._sum_of_squares = 0
        self.multiplier = 1.0 + self._resolution
        self._count = 0
        self._min = self._max_possible
        self._max = 0
        self._buckets = [0] * (self._bucket_for(self._max_possible) + 1)

    def reset(self):
        with self._lock:
            self._sum = 0
            self._sum_of_squares = 0
            self._count = 0
            self._min = self._max_possible
            self._max = 0
            self._buckets = [0] * (self._bucket_for(self._max_possible) + 1)

    def add(self, val):
        with self._lock:
            self._sum += val
            self._sum_of_squares += val * val
            self._count += 1
            self._min = min(self._min, val)
            self._max = max(self._max, val)
            self._buckets[self._bucket_for(val)] += 1

    def get_data(self):
        with self._lock:
            data = stats_pb2.HistogramData()
            data.bucket.extend(self._buckets)
            data.min_seen = self._min
            data.max_seen = self._max
            data.sum = self._sum
            data.sum_of_squares = self._sum_of_squares
            data.count = self._count
            return data

    def _bucket_for(self, val):
        val = min(val, self._max_possible)
        return int(math.log(val, self.multiplier))

File: src/csharp/Grpc.IntegrationTesting/Histogram.cs
#region Copyright notice and license

// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Google.Protobuf;
using Grpc.Core;
using Grpc.Core.Utils;
using NUnit.Framework;
using Grpc.Testing;

namespace Grpc.IntegrationTesting
{
    /// <summary>
    /// Basic implementation of histogram based on grpc/support/histogram.h.
    /// </summary>
    public class Histogram
    {
        readonly object myLock = new object();
        readonly double multiplier;
        readonly double oneOnLogMultiplier;
        readonly double maxPossible;
        readonly uint[] buckets;

        int count;
        double sum;
        double sumOfSquares;
        double min;
        double max;

        public Histogram(double resolution, double maxPossible)
        {
            GrpcPreconditions.CheckArgument(resolution > 0);
            GrpcPreconditions.CheckArgument(maxPossible > 0);
            this.maxPossible = maxPossible;
            this.multiplier = 1.0 + resolution;
            this.oneOnLogMultiplier = 1.0 / Math.Log(1.0 + resolution);
            this.buckets = new uint[FindBucket(maxPossible) + 1];

            ResetUnsafe();
        }

        public void AddObservation(double value)
        {
            lock (myLock)
            {
                AddObservationUnsafe(value);    
            }
        }

        /// <summary>
        /// Gets snapshot of stats and optionally resets the histogram.
        /// </summary>
        public HistogramData GetSnapshot(bool reset = false)
        {
            lock (myLock)
            {
                var histogramData = new HistogramData();
                GetSnapshotUnsafe(histogramData, reset);
                return histogramData;
            }
        }

        /// <summary>
        /// Merges snapshot of stats into <c>mergeTo</c> and optionally resets the histogram.
        /// </summary>
        public void GetSnapshot(HistogramData mergeTo, bool reset)
        {
            lock (myLock)
            {
                GetSnapshotUnsafe(mergeTo, reset);
            }
        }

        /// <summary>
        /// Finds bucket index to which given observation should go.
        /// </summary>
        private int FindBucket(double value)
        {
            value = Math.Max(value, 1.0);
            value = Math.Min(value, this.maxPossible);
            return (int)(Math.Log(value) * oneOnLogMultiplier);
        }

        private void AddObservationUnsafe(double value)
        {
            this.count++;
            this.sum += value;
            this.sumOfSquares += value * value;
            this.min = Math.Min(this.min, value);
            this.max = Math.Max(this.max, value);

            this.buckets[FindBucket(value)]++;
        }

        private void GetSnapshotUnsafe(HistogramData mergeTo, bool reset)
        {
            GrpcPreconditions.CheckArgument(mergeTo.Bucket.Count == 0 || mergeTo.Bucket.Count == buckets.Length);
            if (mergeTo.Count == 0)
            {
                mergeTo.MinSeen = min;
                mergeTo.MaxSeen = max;
            }
            else
            {
                mergeTo.MinSeen = Math.Min(mergeTo.MinSeen, min);
                mergeTo.MaxSeen = Math.Max(mergeTo.MaxSeen, max);
            }
            mergeTo.Count += count;
            mergeTo.Sum += sum;
            mergeTo.SumOfSquares += sumOfSquares;

            if (mergeTo.Bucket.Count == 0)
            {
                mergeTo.Bucket.AddRange(buckets);
            }
            else
            {
                for (int i = 0; i < buckets.Length; i++)
                {
                    mergeTo.Bucket[i] += buckets[i];
                }
            }

            if (reset)
            {
              ResetUnsafe();
            }
        }

        private void ResetUnsafe()
        {
            this.count = 0;
            this.sum = 0;
            this.sumOfSquares = 0;
            this.min = double.PositiveInfinity;
            this.max = double.NegativeInfinity;
            for (int i = 0; i < this.buckets.Length; i++)
            {
                this.buckets[i] = 0;
            }
        }
    }
}
